feat: Add --filename flag to service create command

dsimansk opened the pull request:
## Description

Allow users to create Services from provided YAML or JSON file.

There're couple of open questions:
- Is it necessary to keep `NAME` parameter for the create command?
  - Should the `NAME` parameter take priority over file provided name?
- Should the multi-item (services separated by `---`) YAML be supported as well? 


## Changes

<!-- Please add list of more detailed changes. These changes should be reflected also in the commit messages -->

* Add new flag `--filename` to `service create` command


## Reference

<!-- Please add the corresponding issue number which this pull request is about to fix -->
Fixes #550

<!--
Please add an entrty to CHANGELOG.adoc file, too, as part of your Pull Request.

In the following cases, add a short description of PR to the unreleased section in CHANGELOG.adoc:

- 🎁 New feature
- 🐛 Bug fix
- ✨ Feature Update
- 🐣 Refactoring
- 🗑️ Remove feature or internal logic

See other entries in CHANGELOG.adoc as an example for how to add the entry, including a reference to the corresponding issue/PR

PLEASE DON'T ADD THAT LINE HERE IN THE PULL-REQUEST DESCRIPTION BUT DIRECTLY IN CHANGELOG.ADOC AND ADD CHANGELOG.ADOC AS PART OF YOUR PULL-REQUEST.
-->

<!--
To automatically lint go code in this pull request uncomment the line below. You get feedback as comments on your pull request then -->

<!--
/lint
-->

mattmoor commented on pkg/kn/commands/service/create_test.go:
```diff
@@ -707,3 +710,99 @@ func TestServiceCreateWithClusterLocal(t *testing.T) {
+    "template": {
+		"spec": {
+		"containers": [
+			{ 
```
<!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
			{
```

mattmoor commented on test/e2e/service_file_test.go:
```diff
@@ -0,0 +1,122 @@
+    "template": {
+		"spec": {
+		"containers": [
+			{ 
```
<!--whitespace-bot-->
Remove trailing whitespace:
```suggestion
			{
```

dsimansk commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+		return nil, err
+	}
+	if service.Name != name {
+		return nil, fmt.Errorf("Provided service name '%s' doesn't match name from file '%s'.", name, service.Name)
```
Adressed in https://github.com/knative/client/pull/913/commits/228610a353d1a5487cd0f86f4a688117e38872ca.

maximilien commented:
>Is it necessary to keep NAME parameter for the create command?

Yes. I would suggest just to have it as priority. And if `--file` is passed then ignore it. So the following all work:

`kn service create NAME ...`
`kn service create --file ...`

>Should the NAME parameter take priority over file provided name?

Yes. If both provided use `NAME` and/or error

>Should the multi-item (services separated by ---) YAML be supported as well

Yes. But I can easily see an argument for supporting one service in first pass and adding this after.

maximilien commented on pkg/kn/commands/service/configuration_edit_flags.go:
```diff
@@ -68,6 +68,8 @@ type ConfigurationEditFlags struct {
 	GenerateRevisionName bool
 	ForceCreate          bool
 
+	File string
```
I always debate using `Filename` instead of `File` since there is invariably a `File` object created somewhere later. Of course, `File` is shorter as a flag name so makes sense too... Argh.

dsimansk replied:
Following the `kubectl` example we could settle on `-f, --filename` as `--force` is not using that.

dsimansk replied:
Changed the name in https://github.com/knative/client/pull/913/commits/9a3c1b92bd85e16386ee9613296481877f774cb6.

maximilien commented on test/e2e/service_file_test.go:
```diff
@@ -0,0 +1,122 @@
+	r := test.NewKnRunResultCollector(t, it)
+	defer r.DumpIfFailed()
+
+	tempDir, err := ioutil.TempDir("", "kn-file")
```
Cleanup? Easy with a `defer os.RemoveAll(tempDir)` or similar.

dsimansk replied:
Done in https://github.com/knative/client/pull/913/commits/9a3c1b92bd85e16386ee9613296481877f774cb6.

maximilien commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 512)
+
+	err = decoder.Decode(&service)
+	if err != nil {
+		return nil, err
+	}
```
I can comment these lines and pass all tests. Add a test for bad input file.

dsimansk replied:
Done in https://github.com/knative/client/pull/913/commits/9a3c1b92bd85e16386ee9613296481877f774cb6

maximilien commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+
+	// We need to generate revision to have --force replace working
+	revName, err := servinglib.GenerateRevisionName(editFlags.RevisionName, &service)
+	if err != nil {
+		return nil, err
+	}
```
No tests for problem generating revision as I can comment these and pass tests. This is less important coverage than the comment above.

navidshaikh commented on pkg/kn/commands/service/configuration_edit_flags.go:
```diff
@@ -272,7 +274,8 @@ func (p *ConfigurationEditFlags) AddCreateFlags(command *cobra.Command) {
 	command.Flags().BoolVar(&p.ForceCreate, "force", false,
 		"Create service forcefully, replaces existing service if any.")
-	command.MarkFlagRequired("image")
+	command.Flags().StringVarP(&p.File, "filename", "f", "", "Create service from file.")
```
```suggestion
	command.Flags().StringVarP(&p.File, "filename", "f", "", "Create a service from file.")
```

dsimansk replied:
https://github.com/knative/client/pull/913/commits/d10ecd121af60f0cc2ca1cc8dfea5c797fac7586

navidshaikh commented on pkg/kn/commands/service/configuration_edit_flags.go:
```diff
@@ -68,6 +68,8 @@ type ConfigurationEditFlags struct {
 	GenerateRevisionName bool
 	ForceCreate          bool
 
+	File string
```
```suggestion
	Filename string
```

dsimansk replied:
https://github.com/knative/client/pull/913/commits/d10ecd121af60f0cc2ca1cc8dfea5c797fac7586

navidshaikh commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+		return nil, fmt.Errorf("provided service name '%s' doesn't match name from file '%s'", name, service.Name)
+	}
+	// Set namespace in case it's specified as --namespace
+	service.ObjectMeta.Namespace = namespace
```
should the namespace also follow check as done above for service name ?

rhuss replied:
I think for the namespace we should just allow an override from the CLI. A typical use case is that you create your service maybe for a different environment in the file but want to apply it here in a different context (that's different from the "name" which is an identifier that does not change across environments).

navidshaikh commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+	service.ObjectMeta.Namespace = namespace
+
+	// We need to generate revision to have --force replace working
+	revName, err := servinglib.GenerateRevisionName(editFlags.RevisionName, &service)
```
I am thinking if the client-side revision name generation could be turned off if using files mode?

dsimansk replied:
```
➜  client git:(create-from-file) ✗ cat /tmp/hello.yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
spec:
  template:
    spec:
      containers:
        - image: gcr.io/knative-samples/helloworld-go
          env:
            - name: TARGET
              value: "Go Sample v1"
➜  client git:(create-from-file) ✗ kn service create helloworld-go -f /tmp/hello.yaml 
Creating service 'helloworld-go' in namespace 'default':
  0.028s The Route is still working to reflect the latest desired specification.
  0.061s Configuration "helloworld-go" is waiting for a Revision to become ready.
  0.106s ...
  3.751s ...
  3.795s Ingress has not yet been reconciled.
  3.868s Waiting for load balancer to be ready
  4.190s Ready to serve.
Service 'helloworld-go' created to latest revision 'helloworld-go-5j5mp' is available at URL:
http://helloworld-go.default.example.com
➜  client git:(create-from-file) ✗ kn service create helloworld-go -f /tmp/hello.yaml --force
Replacing service 'helloworld-go' in namespace 'default':
```

When I was testing the flag locally I've experienced the above behaviour. When trying to -recreate the service with `--force` the update hangs and timeouts eventually. However it's not happening for `kn service update` flow or imperative `kn service create` due to the fact that we generate revision name by default. May it a serving bug?

rhuss replied:
yes could be, but we should be sure what it actually happen when using `--force`. Could you do a debug session and check where exactly it hangs ?

For opening an issue for serving we would need to recreate the steps via yaml files, I think.

dsimansk replied:
Well, the issue is limited to a corner case when the same source is used twice, in other words to replace the same already created same Service. In that case such an update op never replaces the original Service content, causing no set of event to be produced in the cluster for wait to finish without timing out.
There's only one synthetic event `ADDED` [received](https://github.com/knative/client/blob/master/pkg/wait/wait_for_ready.go#L179), but nothing else as the server-side doesn't update anything at all.
The same behaviour can be reproduced with `kubectl replace`, meaning that original service isn't actually replaced, is still the same with original timestamp etc., but there's no wait for ready that could be blocking the execution.

That kind of issue isn't present in imperative `create` flow due to the fact that [generated revision name](https://github.com/knative/client/blob/master/pkg/kn/commands/service/configuration_edit_flags.go#L334-L341) is enough of a change to trigger the update/replace mechanism, therefore desired set of events is received in wait function.


I'd suggest the following:

- Implement @rhuss's suggestion from [this comment  ](https://github.com/knative/client/pull/913#discussion_r450685124)
- Mark `filename` flag with `markFlagMakesRevision` to enforce revision generation
  - Adding modification from other flag would always trigger new revision name generation, therefore we'll benefit from it for filename as well (although it might be seen as hack as well)
- => Profit.

rhuss replied:
For a quick win, I would go for the second option. We then can implement that improvement to pick up other options later. wdyt ?

dsimansk replied:
OK, I'm on board with quick win. :) Anyway it'll better to have improvement PR with proper amount of tests and bake time for options combination. I'll create a new issue for it.

dsimansk replied:
Follow-up enhancement issue https://github.com/knative/client/issues/923.

rhuss commented:
> > Is it necessary to keep NAME parameter for the create command?
> 
> Yes. I would suggest just to have it as priority. And if `--file` is passed then ignore it. So the following all work:
> 

I would keep the NAME argument and make it non-mandatory when `--file` is given, too, but I also would never ignore anything that a user provides on the command line as this is not what they expect.

My suggestion:

* no name arg, name in file --> Take name in file
* name arg, no name in file --> Take name arg
* name arg and name in file given, both equal --> Take that name
* name arg and name in file given, but different --> Error

dsimansk commented:
@maximilien @rhuss 
Based on your suggestions I've refactored the initial name param restrictions in commit https://github.com/knative/client/pull/913/commits/d1c8059950ff610955056fa40eceb6f646e0439e.

Please note that I've renamed initial flag name to `-f, --filename` per Michael's suggestion and also to match `kubectl` naming.

Now it's possible to use with [extended conditions](https://github.com/knative/client/pull/913#issuecomment-653435884) suggested by Roland.
```
kn service create --filename ...
kn service create NAME --filename ... 

kn service create -f ..
kn service create NAME -f ... 
```

maximilien commented:
@rhuss will you take a pass? I can do a final review tomorrow after you or if you don’t? I like your suggestion.

rhuss commented on pkg/kn/commands/service/create.go:
```diff
@@ -260,3 +270,42 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
+		return nil, err
+	}
+	service.Spec.Template.Name = revName
+
```
Where does the override from all other options from the CLI happen ? I mean when someone uses

```
kn service create -f mysvc.yml --service-account mysa
```

then `mysa` should be set as service account in the generated service. For array/map like options that's a bit more tricky as we have to decided whether they override those or just work on the ones from the file like with an update (i would opt for the latter). I.e. I would think adding options in addition to the file would act like an update as if the service in the file has been already present on the cluster.

wdyt ?

As this looks a bit more involved I would be happy to add this in a second PR.

dsimansk replied:
Initially I've left it out, because of potential clashes. However from my limited tests so far it looks well good. E.g. env vars are merged, service account added etc. I'll add a few tests to cover those use cases and update the PR. Plus there's another benefit [outlined here](https://github.com/knative/client/pull/913#discussion_r451481222).

maximilien commented:
/hold 
/lgtm

@rhuss feel free to remove hold when ready

maximilien commented:
/unhold

rhuss commented:
/tetest

rhuss commented:
/retest
